build.py

```diff
@@ -515,8 +523,54 @@ def TEMPFILE(cmd_, comstr=None):
             return
 
         tool = str(n.executor).split(None, 1)[0]
-        if tool not in ('$CC', '$CXX', '$SHCC', '$SHCXX', '$LINK', '$SHLINK', '$AR', '$RC'):
-            n.scan() # We need this for IDL.
+        if tool == '$IDLC' and myEnv.get('IDL_HAS_INLINE_DEPENDENCIES'):
+            if n.implicit:
+                implicit_deps += strmap(n.implicit)
+
+            # Important:
+            # Originally we ran the IDL scanner during build.ninja file generation but this is
+            # single-thread and slow as the number of IDL files has grown to greater then 100.
+            # Now, we let IDL tell Ninja its dependencies. Unfortunately, IDL does not have a good
+            # way to this with its implicit dependency caching and multiple outputs due to a
+            # limitation in ninja.
+            # See https://github.com/ninja-build/ninja/pull/1534
+            #
+            # Instead we split IDL file generation into "two" phases:
+            # 1. Generate the header and the cpp file as normal but only tell Ninja about the header
+            # 2. "Generate" the cpp file by telling Ninja it depends on the header file via a
+            #    phony rule
+            #
+            idl_cpp_file = strmap(targets)[0]
+            idl_header_file = strmap(targets)[1]
+
+            # Append --write-dependencies-inline so that IDL generates a list of dependencies when
+            # it runs
+            idl_command = self.make_command(myEnv.subst(str(n.executor), executor=n.executor) \
+                + " --write-dependencies-inline")
+
+            implicit_deps.append("_generated_idl")
```

Comment:
This is the _idl_generator, not generated idl.

But more importantly, there is a downside of this form of dependency that I forgot. If one of those files is deleted, it will fail to build because it thinks it needs that file. This is avoided right now by the fact the build.ninja is rebuild whenever any .py file under buildscripts/ is modified, using the depsfile-style dependencies, but it may be nice to someday make build.ninja not depend on the IDL and resmoke files, among others. How hard would it be to emit these files along with the .idl files inside of idlc.py? If it'll be a pain, I can merge this now.

Reply:
I would not worry about them being deleted. I have no made the problem any worse as the build script already generated implicit dependencies on all the python scripts.

I would rather not make the idl compiler generate the python dependencies because then it has to understand where it is located in the build tree to generate the right list and that causes a lot of aggravation. 

See https://github.com/mongodb/mongo/blob/master/site_scons/site_tools/idl_tool.py#L54-L56